scripts/output_env.sh

```diff
@@ -41,81 +45,100 @@ print_version()
+echo "\n"
+echo "Tool Versions:\n"
+
 if [ "${RUN_ARMCC:-1}" -ne 0 ]; then
     : "${ARMC5_CC:=armcc}"
-    print_version "$ARMC5_CC" "--vsn" "armcc not found!" "head -n 2"
+    print_version "$ARMC5_CC" "--vsn" "" "head -n 2"
     echo
 
     : "${ARMC6_CC:=armclang}"
-    print_version "$ARMC6_CC" "--vsn" "armclang not found!" "head -n 2"
+    print_version "$ARMC6_CC" "--vsn" "" "head -n 2"
     echo
 fi
 
-print_version "arm-none-eabi-gcc" "--version" "gcc-arm not found!" "head -n 1"
+print_version "arm-none-eabi-gcc" "--version" "" "head -n 1"
 echo
 
-print_version "gcc" "--version" "gcc not found!" "head -n 1"
+print_version "gcc" "--version" "" "head -n 1"
 echo
 
-print_version "clang" "--version" "clang not found" "head -n 2"
+print_version "clang" "--version" "" "head -n 2"
 echo
 
-print_version "ldd" "--version"                     \
-    "No ldd present: can't determine libc version!" \
-    "head -n 1"
+print_version "ldd" "--version" "" "head -n 1"
 echo
 
-print_version "valgrind" "--version" "valgrind not found!"
+print_version "valgrind" "--version" ""
+echo
+
+print_version "gdb" "--version" "" "head -n 1"
 echo
 
 : ${OPENSSL:=openssl}
-print_version "$OPENSSL" "version" "openssl not found!"
+print_version "$OPENSSL" "version" "default"
 echo
 
 if [ -n "${OPENSSL_LEGACY+set}" ]; then
-    print_version "$OPENSSL_LEGACY" "version" "openssl legacy version not found!"
-    echo
+    print_version "$OPENSSL_LEGACY" "version" "legacy"
+else
+    echo " * openssl (legacy): Not configured."
 fi
+echo
 
 if [ -n "${OPENSSL_NEXT+set}" ]; then
-    print_version "$OPENSSL_NEXT" "version" "openssl next version not found!"
-    echo
+    print_version "$OPENSSL_NEXT" "version" "next"
+else
+    echo " * openssl (next): Not configured."
 fi
+echo
 
 : ${GNUTLS_CLI:=gnutls-cli}
-print_version "$GNUTLS_CLI" "--version" "gnuTLS client not found!" "head -n 1"
+print_version "$GNUTLS_CLI" "--version" "default" "head -n 1"
 echo
 
 : ${GNUTLS_SERV:=gnutls-serv}
-print_version "$GNUTLS_SERV" "--version" "gnuTLS server not found!" "head -n 1"
+print_version "$GNUTLS_SERV" "--version" "default" "head -n 1"
 echo
 
 if [ -n "${GNUTLS_LEGACY_CLI+set}" ]; then
-    print_version "$GNUTLS_LEGACY_CLI" "--version" \
-        "gnuTLS client legacy version not found!"  \
-        "head -n 1"
-    echo
+    print_version "$GNUTLS_LEGACY_CLI" "--version" "legacy" "head -n 1"
+else
+     echo " * gnutls-cli (legacy): Not configured."
 fi
+echo
 
 if [ -n "${GNUTLS_LEGACY_SERV+set}" ]; then
-    print_version "$GNUTLS_LEGACY_SERV" "--version" \
-        "gnuTLS server legacy version not found!"   \
-        "head -n 1"
-    echo
+    print_version "$GNUTLS_LEGACY_SERV" "--version" "legacy" "head -n 1"
+else
+    echo " * gnutls-serv (legacy): Not configured."
 fi
+echo
 
-if `hash dpkg > /dev/null 2>&1`; then
-    echo "* asan:"
+echo " * Installed asan versions:"
+if `hash dpkg > /dev/null 2>&1` && `dpkg --get-selections|grep libasan > /dev/null` ; then
```

Comment:
`dpkg --get-selections` is technically wrong, although it won't matter much in practice: `--get-selections` is what the administrator wants to install, which might not match the current state of the system. You can get a list of packages with `dpkg-query -f '${Status} ${Package}\n' -W 'libasan*' | awk '$3 == "installed" && $4 ~! /-/ {print $4}'` (bonus: it'll still work with `libasan10`). And with this approach, you can list the versions as well (including libasan6 and so on):

```
if type dpkg-query >/dev/null 2>/dev/null; then
    if ! dpkg-query -f '${Status} ${Package}: ${Version}\n' -W 'libasan*' |
         awk '$3 == "installed" && $4 !~ /-/ {print $4, $5}' |
         grep .
    then
        echo "   No asan versions installed."
else
    echo "  Unable to determine the asan version without dpkg."
fi
```

Reply:
I just used this as it looks neater and smarter, although I added a missing `fi`.

However, what is the `grep .` for? This looks like an idiom I don't know, and I can't see what difference it makes.

Reply:
I think the goal of `grep .` is to only select non-empty lines. I personally would find the spelling `grep -v '^$'` clearer, but perhaps it's just because I'm not used to that idiom. (It's only the second time I see it, and the first time I misunderstood the intent so Gilles had to clarify.)

Reply:
Aha, interestingly enough, after reviewing the code, I now think the purpose of the `| grep .` here is what I thought it was the first time I saw it (in another PR): not so much to select non-empty lines, but more to return success if and only if the output of the previous command was non-empty (which is to say, it had at least one non-empty line). I just checked the man page and `awk` returns success if no error occurred, which is fair enough but not what we want here: we want to detect if any version of libasan was found.

Reply:
Thanks for the explanation @mpg!

Reply:
Here's how the pipeline works:

1. List `libasan*` packages with certain metadata.
2. Filter only installed packages whose name doesn't contain a dash, and print the package name and version.
3. Test if there is at least one matching package (without suppressing the output).

To me step 3 is idiomatic in context, but feel free to add a comment or separate out the steps.

Looking back I'm actually not so happy about step 2 without documentation. I normally assume that the reader is familiar with the basics of the programming language, and this is simple awk, but to understand what's going on, you need to know what information the fields `$3`, `$4` and `$5` contain. This depends on the fairly obscure knowledge of the format of `dpkg-query`, and in particular of the gotcha that `${Status}` is three fields, not one, which is not even documented in the `dpkg-query` man page. So I suggest

```
         # List libasan* packages as "DESIRED_STATUS ERROR_STATUS ACTUAL_STATUS NAME VERSION"
    if ! dpkg-query -f '${Status} ${Package}: ${Version}\n' -W 'libasan*' |
         # Keep only installed packages without a "-" in their name
         # (remove e.g. *-dbg and *-cross)
         awk '$3 == "installed" && $4 !~ /-/ {print $4, $5}' |
         # Test if there is at least one match
         grep .
    then
        echo "   No asan versions installed."
    fi
```

or

```
    libasan_packages=$(
        # List libasan* packages as "DESIRED_STATUS ERROR_STATUS ACTUAL_STATUS NAME VERSION"
        dpkg-query -f '${Status} ${Package}: ${Version}\n' -W 'libasan*' |
        # Keep only installed packages without a "-" in their name
        # (remove e.g. *-dbg and *-cross)
        awk '$3 == "installed" && $4 !~ /-/ {print $4, $5}'
    )
    if [ -n "$libasan_packages" ]; then
        echo "$libasan_packages"
    else
        echo "   No asan versions installed."
    fi
```